distributed/dashboard/scheduler.py

```diff
@@ -253,40 +258,53 @@ def collect(self):
             value=self.server.adaptive_target(),
         )
 
-        tasks = GaugeMetricFamily(
+        worker_states = GaugeMetricFamily(
             "dask_scheduler_workers",
             "Number of workers known by scheduler.",
             labels=["state"],
         )
-        tasks.add_metric(["connected"], len(self.server.workers))
-        tasks.add_metric(["saturated"], len(self.server.saturated))
-        tasks.add_metric(["idle"], len(self.server.idle))
-        yield tasks
+        worker_states.add_metric(["connected"], len(self.server.workers))
+        worker_states.add_metric(["saturated"], len(self.server.saturated))
+        worker_states.add_metric(["idle"], len(self.server.idle))
+        yield worker_states
 
         tasks = GaugeMetricFamily(
             "dask_scheduler_tasks",
             "Number of tasks known by scheduler.",
             labels=["state"],
         )
-        tasks.add_metric(["received"], len(self.server.tasks))
-        tasks.add_metric(["unrunnable"], len(self.server.unrunnable))
+
+        task_counter = merge_with(
+            sum, (tp.states for tp in self.server.task_prefixes.values())
+        )
+
+        yield CounterMetricFamily(
+            "dask_scheduler_tasks_forgotten",
+            "Number of tasks already forgotten by scheduler.",
+            value=task_counter.get("forgotten", 0.0),
+        )
+
+        for state in ALL_TASK_STATES:
+            tasks.add_metric([state], task_counter.get(state, 0.0))
         yield tasks
 
 
 class PrometheusHandler(RequestHandler):
-    _initialized = False
+    _collector = None
 
     def __init__(self, *args, **kwargs):
         import prometheus_client
 
         super(PrometheusHandler, self).__init__(*args, **kwargs)
 
-        if PrometheusHandler._initialized:
+        if PrometheusHandler._collector:
+            # Especially during testing, multiple schedulers are started
+            # sequentially in the same python process
+            PrometheusHandler._collector.server = self.server
```

Comment:
This is why the tests didn't run on travis. I only ran one of the prometheus tests on my machine but travis obviously executes all. The first test would register a collector with its scheduler. Every subsequent test would then receive the request handler for the scheduler of the first test instead of the current one.

This is a rather pragmatic fix since I expect these situations not to actually happen in real scenarios and I didn't want to introduce anything too complex but I'm open for suggestions.